### PR TITLE
fix(app): collision-resistant group keys in insight grouping

### PR DESCRIPTION
**Pull request by youhaowei**
Replace the delimiter-join group key strategy in `groupRowsBy` with `JSON.stringify` over typed per-field tuples. This prevents distinct row value-combinations from being merged into the same group when a field value contains the old `|||` delimiter or the strings `"null"`/`"__NULL__"`.

## Changes

- **`groupRowsBy`** (`compute-preview.ts`): replace `keyParts.join("|||")` with `JSON.stringify` over `["null"]` | `["v", value]` typed tuples per field. Each field now uses `field.columnName ?? field.id` as the row lookup key so fields without `columnName` participate in grouping instead of silently collapsing to null.
- **`extractGroupKeys`** (`compute-preview.ts`): same `field.columnName ?? field.id` fallback so virtual/computed fields appear in the output row (previously silently dropped).
- **`compute-preview.test.ts`** (new): 8 tests covering delimiter collision, null/sentinel collision, columnName-absent participation, and clean-value regression.

## Screenshots

No UI change — compute correctness fix.

## Verification

- `bun run test` in `packages/app`: 404 tests pass (8 new)
- `bun run typecheck`: clean across all packages
- `bun run lint`: clean
- `bun run format:check`: clean
- `bun run check:ticket-refs`: no ticket refs in source

<details open><summary><h3>Greptile Summary</h3></summary>

This PR replaces the fragile `|||`-delimiter-join group key strategy in `groupRowsBy` and `extractGroupKeys` with typed tuples serialized via `JSON.stringify`, eliminating delimiter collisions, null/sentinel collisions, and the silent dropping of fields without `columnName`.

- **`groupRowsBy`**: key encoding now dispatches on runtime type — `[\"null\"]` for null/undefined, `[\"num\", str]` for non-finite numbers, `[\"bigint\", str]` for BigInt, and `[\"v\", value]` for everything else — and uses `field.columnName ?? field.name` for row lookup so virtual/computed fields participate in grouping.
- **`extractGroupKeys`**: adopts the same `columnName ?? name` fallback so virtual fields are no longer silently omitted from the output row.
- **`compute-preview.test.ts`** (new): eight focused tests cover delimiter collision, null/sentinel distinction, undefined handling, NaN/±Infinity separation, BigInt grouping, cross-type isolation, and regression on clean values.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the key encoding is correct for every value type DuckDB produces, the NaN/Infinity edge case is properly handled, and all eight new tests pass.

The typed-tuple encoding is injective across all cases (null, NaN, ±Infinity, BigInt, string, finite number, boolean), no existing behavior is regressed, and the new tests directly exercise every former failure mode including the previously flagged NaN/Infinity collapse.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/app/src/lib/insights/compute-preview.ts | Replaces the delimiter-join key strategy with typed tuple encoding; handles all DuckDB-producible value types (null, NaN, Infinity, BigInt, string, number, boolean) without collision or throw. Logic is correct. |
| packages/app/src/lib/insights/compute-preview.test.ts | New test file with eight tests covering every fixed invariant: delimiter collision, null/sentinel/undefined distinction, NaN and ±Infinity isolation, BigInt grouping and cross-type separation, columnName-absent participation, and grand-total mode regression. |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Row value] --> B{value == null or undefined?}
    B -- Yes --> C[tag: null]
    B -- No --> D{typeof value === bigint?}
    D -- Yes --> E[tag: bigint str]
    D -- No --> F{typeof value === number and not isFinite?}
    F -- Yes --> G[tag: num str - NaN Infinity -Infinity]
    F -- No --> H[tag: v value - string finite number boolean]
    C --> I[JSON.stringify all field tags to group key]
    E --> I
    G --> I
    H --> I
    I --> J{Key in groupMap?}
    J -- No --> K[Create new group]
    J -- Yes --> L[Append row to existing group]
```
</details>

<sub>Reviews (3): Last reviewed commit: ["fix(app): total injective group key enco..."](https://github.com/youhaowei/dashframe/commit/6e8c4f995efdc98e75504577144b899e1116153a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37447431)</sub>

### CODE REVIEW

File: packages/app/src/lib/insights/compute-preview.test.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+// Helpers
+// ---------------------------------------------------------------------------
+
+function field(overrides: Partial<Field> & Pick<Field, "id" | "name">): Field {
+  return {
+    tableId: "t1",
+    type: "string",
+    ...overrides,
+  } as Field;
+}
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/lib/insights/compute-preview.test.ts:37]: Fixed in 8b3cc628: explicitly listing all required Field properties (id, name, tableId, type) in the object literal so TypeScript validates the return type structurally without the cast.

File: packages/app/src/lib/insights/compute-preview.ts [hunk 2]
```diff
@@ -122,12 +122,19 @@ function groupRowsBy(
...
+    // collapse to the same null bucket.
     const keyParts = fields.map((field) => {
-      const value = field.columnName ? row[field.columnName] : null;
-      return value != null ? String(value) : "__NULL__";
+      const colKey = field.columnName ?? field.id;
+      const value = row[colKey];
+      return value == null ? ["null"] : ["v", value];
     });
-    const key = keyParts.join("|||");
+    const key = JSON.stringify(keyParts);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/lib/insights/compute-preview.ts:152]: Fixed in 9bfbeedf: added a ['special', String(value)] tag for non-finite numbers before JSON.stringify so NaN, Infinity, and -Infinity each produce a distinct key and don't collapse to the same 'null' as actual null values. Added a test covering all four variants.

[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/lib/insights/compute-preview.ts:152]: Addressed in 9bfbeedf: non-finite numbers (NaN, Infinity, -Infinity) are now serialized with a ['special', String(value)] tag before JSON.stringify, preventing them from collapsing to null.

[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/lib/insights/compute-preview.ts:152]: Also addressed in 6e8c4f99: BigInt values (DuckDB BIGINT columns surface as JS bigint) now get a ["bigint", val.toString()] tag before JSON.stringify — the previous commit would have thrown TypeError on any BIGINT-keyed group. The columnName fallback was also corrected from field.id to field.name to match the codebase convention.

**Comment by youhaowei** [comment]
Re: the bigint + non-finite group-key threads — these are addressed in the current head (6e8c4f99). compute-preview.ts:146-148 intercepts `bigint` → `["bigint", value.toString()]` and non-finite numbers → `["num", String(value)]` BEFORE the `["v", value]` branch, so `JSON.stringify` never receives a raw bigint (no throw) and Infinity/-Infinity/NaN each get a distinct tagged key (no collapse). Tests cover all four non-finite cases as distinct groups + bigint cross-type distinctness. The threads referenced `[["v", 1n]]`/`[["v", Infinity]]` which the code no longer produces — stale analysis on a superseded commit. Resolving.
